### PR TITLE
Make default projection of extents configurable

### PR DESCRIPTION
**Pull request by Vampouille**

### CODE REVIEW

**Comment by Vampouille** [comment]
WIP please do not merge !

**Comment by Vampouille** [comment]
Il faut convertir en WGS 84 la table `non_validated` et peut être d'autres.
```sql
ALTER TABLE non_validated ALTER COLUMN the_geom TYPE geometry(MULTIPOLYGON, 4326) USING ST_Transform(ST_SetSRID(the_geom,21781),4326) ;
```

File: web-ui/src/main/resources/catalog/geocat-shared-objects/js/extent-directive.js [hunk 1]
```diff
@@ -94,9 +94,10 @@
                   return;
                 }
                 var geom = feature.getGeometry().clone();
-                if (!scope.incompatible) {
+
+                /*if (!scope.incompatible) {
                   geom = geom.transform(map.getView().getProjection(), scope.formObj.proj);
-                }
+                }*/
```
Comment (Vampouille) [inline web-ui/src/main/resources/catalog/geocat-shared-objects/js/extent-directive.js]: Il y a aura peut être un effet de bord ... à tester

**Comment by Vampouille** [comment]
```sql
ALTER TABLE xlinks DROP CONSTRAINT enforce_srid_the_geom;

ALTER TABLE xlinks ALTER COLUMN the_geom TYPE geometry(MULTIPOLYGON, 4326) USING ST_Transform(ST_SetSRID(the_geom,21781),4326);
```

**Comment by Vampouille** [comment]
J'ai regardé l'état de la base de donnée au niveau postgis. J'ai l'impression qu'il y a un mix de postgis 1.5 et 2 avec des contraintes sur les tables et dans la vue `geometry_columns`.
Par exemple ici : 
https://github.com/camptocamp/geocat/blob/geocat_develop/web/src/main/webapp/WEB-INF/classes/setup/sql/data/create-db-geocat-default.sql#L13-L15
Avec postgis 2, on peut utiliser directement ce type : `geometry(MULTIPOLYGON, 4326)`
Cette requête par exemple :
https://github.com/camptocamp/geocat/blob/geocat_develop/web/src/main/webapp/WEB-INF/classes/setup/sql/data/create-db-geocat-default.sql#L369-L371
ne fonctionne plus sous postgis 2

**Comment by fgravin** [comment]
> . J'ai l'impression qu'il y a un mix de postgis 1.5 et 2

C'est tout à fait probable.

**Comment by Vampouille** [comment]
La liste déroulante sur l'interface d'édition est maintenant configuré par le paramètre dans `config-spring-geocat.xml`

**Comment by Vampouille** [comment]
In order to switch from 21781 to 4326 you need to : 

```sql
ALTER TABLE public.xlinks ALTER COLUMN the_geom TYPE geometry(MultiPolygon, 4326) USING ST_Transform(the_geom, 4326);
ALTER TABLE public.non_validated ALTER COLUMN the_geom TYPE geometry(MultiPolygon, 4326) USING ST_Transform(the_geom, 4326);
```
And edit `web/src/main/webapp/WEB-INF/config-spring-geocat.xml` change projection of following beans :
* gn:xlinks
* gn:non_validated

Finally, change Region Map API in geocat settings.

File: services/src/main/java/org/fao/geonet/services/region/GetMap.java [hunk 2]
```diff
@@ -180,6 +181,11 @@ public void init() {
                                    @RequestParam(value = OUTPUT_FILE_NAME, required=false) String outputFileName,
                                    NativeWebRequest request) throws Exception {
 
+
+        // force getMap to use configured projection
+        FeatureType nonValidated = context.getBean("gn:non_validated", FeatureType.class);
+        srs = nonValidated.srs();
```
Comment (pvalsecc) [inline services/src/main/java/org/fao/geonet/services/region/GetMap.java]: Shouldn't we remove the param, then?

File: web-ui/src/main/resources/catalog/geocat-shared-objects/js/extent-directive.js [hunk 3]
```diff
@@ -211,17 +220,36 @@
...
+                  // Reproject geometry to map projection
+                  f.setGeometry(geom.transform(gnGlobalSettings.srs, scope.map.getView().getProjection()));
+                  featureOverlay.addFeature(f);
+
                 }
 
                 setTimeout(function () {
                   scope.map.updateSize();
+                  if(featureType && featureType.feature.geom) {
+                    console.log(scope.map.getSize());
```
Comment (pvalsecc) [inline web-ui/src/main/resources/catalog/geocat-shared-objects/js/extent-directive.js]: You sure you want to keep that line?

File: web-ui/src/main/resources/catalog/geocat-shared-objects/partials/extent.html [hunk 4]
```diff
@@ -31,8 +31,8 @@
     <label class="col-sm-2 control-label" translate>projection</label>
     <div class="col-sm-4">
       <select class="form-control input-sm" data-ng-model="formObj.proj" ng-change="fillInput()">
-        <option value="EPSG:21781" translate>CHO3 (EPSG:21781)</option>
-        <option value="EPSG:4326" translate>WGS 84 (EPSG:4326)</option>
+        <option value="EPSG:4326" translate ng-if="gnGlobalSettings.srs == 'EPSG:4326'">WGS 84 (EPSG:4326)</option>
+        <option value="EPSG:21781" translate ng-if="gnGlobalSettings.srs == 'EPSG:21781'">CHO3 (EPSG:21781)</option>
       </select>
```
Comment (pvalsecc) [inline web-ui/src/main/resources/catalog/geocat-shared-objects/partials/extent.html]: What's the point of having a select if it has only one possible value?

File: web-ui/src/main/resources/catalog/views/geocat/config.js [hunk 5]
```diff
@@ -43,52 +43,113 @@
...
-      ];
-
-      var matrixIds = [];
-      for (var i = 0; i < resolutions.length; i++) {
-        matrixIds.push(i);
+      if(gnGlobalSettings.srs == "EPSG:4326" ) {
+
+         // Configuration for 4326
+
+         var projection = ol.proj.get('EPSG:3857');
```
Comment (pvalsecc) [inline web-ui/src/main/resources/catalog/views/geocat/config.js]: What is EPSG:3857? Why not 4326?

File: web-ui/src/main/resources/catalog/views/geocat/config.js [hunk 6]
```diff
@@ -43,52 +43,113 @@
...
+         var projectionExtent = projection.getExtent();
+         var size = ol.extent.getWidth(projectionExtent) / 256;
+         var resolutions = new Array(14);
+         var matrixIds = new Array(14);
+         for (var z = 0; z < 14; ++z) {
+         // generate resolutions and matrixIds arrays for this WMTS
+         resolutions[z] = size / Math.pow(2, z);
+         matrixIds[z] = z;
```
Comment (pvalsecc) [inline web-ui/src/main/resources/catalog/views/geocat/config.js]: Indentation

File: web-ui/src/main/resources/catalog/views/geocat/editor/SharedObjectDirective.js [hunk 7]
```diff
@@ -314,7 +316,21 @@
...
+        var serverProj = gnGlobalSettings.srs;
+
+        var formObj = angular.copy(scope.formObj);
+        var formatWKT = new ol.format.WKT();
+
+        if (formObj.proj != serverProj) {
+          formObj.geomString = formatWKT.writeGeometry(formatWKT.readGeometry(formObj.geomString).transform(formObj.proj, serverProj));
+          formObj.proj = serverProj;
+        }
+        */
```
Comment (pvalsecc) [inline web-ui/src/main/resources/catalog/views/geocat/editor/SharedObjectDirective.js]: Do you want to keep this comment?

File: web/src/main/webapp/WEB-INF/config-spring-geocat.xml [hunk 8]
```diff
@@ -67,7 +67,7 @@
         <property name="descColumn" value="DESC"/>
         <property name="searchColumn" value="SEARCH"/>
         <property name="geoIdColumn" value="GEO_ID"/>
-        <property name="srs" value="EPSG:21781"/>
+        <property name="srs" value="EPSG:4326"/>
```
Comment (pvalsecc) [inline web/src/main/webapp/WEB-INF/config-spring-geocat.xml]: I'm not sure I understand the implications of that, but it looks suspicious...

File: services/src/main/java/org/fao/geonet/services/region/GetMap.java (hunk 2 above)
Comment (Vampouille) [inline services/src/main/java/org/fao/geonet/services/region/GetMap.java]: yes, or find groovy template that generate link

File: web-ui/src/main/resources/catalog/geocat-shared-objects/js/extent-directive.js (hunk 3 above)
Comment (Vampouille) [inline web-ui/src/main/resources/catalog/geocat-shared-objects/js/extent-directive.js]: you're right ! I can remove this

File: web-ui/src/main/resources/catalog/views/geocat/config.js (hunk 5 above)
Comment (Vampouille) [inline web-ui/src/main/resources/catalog/views/geocat/config.js]: This is projection of base layer used by https://server.arcgisonline.com/arcgis/rest/services/World_Topo....

File: web-ui/src/main/resources/catalog/views/geocat/config.js (hunk 6 above)
Comment (Vampouille) [inline web-ui/src/main/resources/catalog/views/geocat/config.js]: ok

File: web-ui/src/main/resources/catalog/geocat-shared-objects/partials/extent.html (hunk 4 above)
Comment (Vampouille) [inline web-ui/src/main/resources/catalog/geocat-shared-objects/partials/extent.html]: I just want to keep original layout

File: web/src/main/webapp/WEB-INF/config-spring-geocat.xml [hunk 9]
```diff
@@ -76,7 +76,7 @@
...
```
Comment (Vampouille) [inline web/src/main/webapp/WEB-INF/config-spring-geocat.xml]: This value is used to configure geocat default projection.

File: web/src/main/webapp/WEB-INF/config-spring-geocat.xml (hunk 8 above)
Comment (Vampouille) [inline web/src/main/webapp/WEB-INF/config-spring-geocat.xml]: This is used when geocat fetch geometry from db.

Comment (Vampouille) [inline web/src/main/webapp/WEB-INF/config-spring-geocat.xml]: This is how to configure geocat to handle world wide geometry (purpose of this PR).

**Comment by Vampouille** [comment]
En cas d'erreur : `ERROR:  type modifier is not allowed for type "geometry"`
```sql
ALTER TABLE xlinks DROP CONSTRAINT enforce_srid_the_geom;
UPDATE xlinks SET the_geom = ST_Transform(the_geom, 4326);
ALTER TABLE xlinks
  ADD CONSTRAINT enforce_srid_the_geom CHECK (srid(the_geom) = 4326);

ALTER TABLE non_validated DROP CONSTRAINT enforce_srid_the_geom;
UPDATE non_validated SET the_geom = ST_Transform(the_geom, 4326);
ALTER TABLE non_validated
  ADD CONSTRAINT enforce_srid_the_geom CHECK (srid(the_geom) = 4326);
```

**Comment by Vampouille** [comment]
> ALTER TABLE xlinks DROP CONSTRAINT enforce_srid_the_geom;
> 
> ALTER TABLE xlinks ALTER COLUMN the_geom TYPE geometry(MULTIPOLYGON, 4326) USING ST_Transform(ST_SetSRID(the_geom,21781),4326);

Dans le cas d'une base de données en postgis 1.5, il faut modifier la table `geometry_columns`

**Comment by Vampouille** [comment]
@fgravin @pmauduit est-ce que l'un d'entre vous peut faire une review sur cette PR ?

File: core/src/main/java/org/fao/geonet/util/XslUtil.java [hunk 10]
```diff
@@ -135,6 +139,37 @@ public static String getSettingValue(String key) {
...
+
+        ServletContext servletContext = serviceContext.getServlet().getServletContext();
+        ConfigurableApplicationContext springContext = JeevesDelegatingFilterProxy.getApplicationContextFromServletContext(servletContext);
+
+        Object o = springContext.getBean(beanId);
+        Field f = o.getClass().getDeclaredField(propName);
+        boolean isAccessible = f.isAccessible();
+        f.setAccessible(true);
+        String res = f.get(o).toString();
+        f.setAccessible(isAccessible);
```
Comment (pmauduit) [inline core/src/main/java/org/fao/geonet/util/XslUtil.java]: I wonder if this is useful to setAccessible() back to the original value. I tried the following groovy snippet:

```groovy
import java.lang.reflect.Field;
class Blah {
 private String field;
}
def b = new Blah();
Field f = b.getClass().getDeclaredField("field");
println f.isAccessible();

f.setAccessible(true);

f= b.getClass().getDeclaredField("field");
println f.isAccessible();

```
Which outputs the following:

```
false
false
```

My guess is that each time you will call getDeclaredField(), you will get a new object with the original "access state" of your members in your class. Anyway, this just argues that the line 169 can be removed, it does not harm though.

**Comment by fgravin** [comment]
Don't have much to say about the js code, too hard to review outside from context.
Looks pertty Ok.

But i don't understand the complexity of all the bean property thing ... everything could have been done way more easly with DB param.

File: core/src/main/java/org/fao/geonet/util/XslUtil.java (hunk 10 above)
Comment (Vampouille) [inline core/src/main/java/org/fao/geonet/util/XslUtil.java]: Agree to remove L169

**Comment by pmauduit** [comment]
After having asked @Vampouille and @fgravin, it should be safe to be merged.

**Pull request merged**